src/GUI/GameMoveListener.java

```diff
@@ -0,0 +1,115 @@
+package GUI;
+
+import Cards.Card;
+import Gameplay.*;
+
+import javax.swing.event.MouseInputAdapter;
+import java.awt.*;
+import java.awt.event.MouseEvent;
+
+public class GameMoveListener extends MouseInputAdapter {
+
+    private Deck deck = GamePanel.getDeck();
+    private WastePile wastePile = null;
+    private Tableau selectedTableau = null;
+    private Foundations selectedFoundation = null;
+    private Card selectedCard = null;
+    private ShuffleButton selectedShuffleButton = null;
+
+
+    @Override
+    public void mouseClicked(MouseEvent mouseEvent) {
+        Component pressedComponent = mouseEvent.getComponent().getComponentAt(mouseEvent.getPoint());
+
+        if(pressedComponent instanceof Foundations) {
+            selectedShuffleButton =null;
+            selectedFoundation = (Foundations) pressedComponent;
+            selectedTableau = null;
+            wastePile = null;
+            selectedCard = selectedFoundation.topCard();
+        }else if (pressedComponent instanceof Tableau) {
+            selectedTableau = (Tableau) pressedComponent;
+            wastePile = null;
+            selectedShuffleButton = null;
+            selectedCard = selectedTableau.getClickedCard(mouseEvent.getY() - 100);
+            System.out.println("Click y coordinate: " + mouseEvent.getY());
+            System.out.println("Selected card from tableau: " + selectedCard);
+
+            for(Foundations foundations: GamePanel.getFoundations()) {
+                if(selectedTableau.moveTo(foundations, selectedCard)){
```

Comment:
I believe I had a bug due to this. 

But TL;DR, try to avoid having side-effects in conditionals.
- Conditions should not alter the impacted objects. 
As in, the minimum - first do the `moveTo`, then ask what the result is (cleaner).
Better though, have a separate `ifValid` function, then `moveToUnchecked`